drivers/regulator/regulator_axp192.c

```diff
@@ -0,0 +1,421 @@
+	.ranges = ldo3_ranges,
+	.num_ranges = ARRAY_SIZE(ldo3_ranges),
+};
+
+int axp192_enable(const struct device *dev)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	const struct regulator_axp192_common_config *cconfig = config->parent->config;
+	int ret = 0;
+
+	LOG_INST_DBG(config->log, "Enabling regulator");
+	LOG_INST_DBG(config->log, "[0x%02x]=0x%02x mask=0x%02x", (uint32_t)config->desc->enable_reg,
+		     (uint32_t)config->desc->enable_val, (uint32_t)config->desc->enable_mask);
+
+	ret = i2c_reg_update_byte_dt(&cconfig->i2c, config->desc->enable_reg,
+				     config->desc->enable_mask, config->desc->enable_val);
+
+	if (0 != ret) {
+		LOG_INST_ERR(config->log, "Failed to enable regulator");
+	}
+
+	return ret;
+}
+
+int axp192_disable(const struct device *dev)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	const struct regulator_axp192_common_config *cconfig = config->parent->config;
+	int ret = 0;
+
+	LOG_INST_DBG(config->log, "Disabling regulator");
+	LOG_INST_DBG(config->log, "[0x%02x]=0 mask=0x%x", (uint32_t)config->desc->enable_reg,
+		     (uint32_t)config->desc->enable_mask);
+
+	ret = i2c_reg_update_byte_dt(&cconfig->i2c, config->desc->enable_reg,
+				     config->desc->enable_mask, 0u);
+
+	if (0 != ret) {
+		LOG_INST_ERR(config->log, "Failed to disable regulator");
+	}
+
+	return ret;
+}
+
+unsigned int axp192_count_voltages(const struct device *dev)
+{
+	const struct regulator_axp192_config *config = dev->config;
+
+	return linear_range_group_values_count(config->desc->ranges, config->desc->num_ranges);
+}
+
+int axp192_list_voltage(const struct device *dev, unsigned int idx, int32_t *volt_uv)
+{
+	const struct regulator_axp192_config *config = dev->config;
+
+	return linear_range_group_get_value(config->desc->ranges, config->desc->num_ranges, idx,
+					    volt_uv);
+}
+
+int axp192_set_voltage(const struct device *dev, int32_t min_uv, int32_t max_uv)
+{
+	const struct regulator_axp192_config *config = dev->config;
+	const struct regulator_axp192_common_config *cconfig = config->parent->config;
+	uint16_t idx = 0;
+	int ret = 0;
+
+	ret = linear_range_group_get_win_index(config->desc->ranges, config->desc->num_ranges,
+					       min_uv, max_uv, &idx);
+	if (ret == -EINVAL) {
+		LOG_INST_ERR(config->log, "No voltage range window could be deteceted");
+		return ret;
+	}
+
+	idx <<= config->desc->vsel_bitpos;
+
+	LOG_INST_DBG(config->log, "Setting voltage=%d", config->modes_uv);
+	LOG_INST_DBG(config->log, "[0x%x]=0x%x mask=0x%x", (uint32_t)config->desc->vsel_reg, idx,
+		     (uint32_t)config->desc->vsel_mask);
+	ret = i2c_reg_update_byte_dt(&cconfig->i2c, config->desc->vsel_reg, config->desc->vsel_mask,
+				     (uint8_t)idx);
+
+	if (0 != ret) {
+		LOG_INST_ERR(config->log, "Failed to set regulator voltage");
+	}
+
+	return ret;
+}
+
+int axp192_get_voltage(const struct device *dev, int32_t *volt_uv)
+{
+	const struct regulator_axp192_config *config = dev->config;
```

Comment:
It may be worth considering refactoring this to the mfd approach:
https://github.com/zephyrproject-rtos/zephyr/pull/58229

Reply:
I added the mfd layer. As of now the mdd driver does not feature any additional functionality expect to detect if the chip is available. Further functionality may follow with adding add, gpio and fuel gauge drivers. I think the mdd layer could verify consistency of individual settings and especially pinctrl. But that's something I'll have a separate look at.
For now there wasn't any benefit or need to move the i2c communication to mfd layer, as the standard i2c driver interface fulfills all needs, included atomic read-modify-write operations.

Reply:
Maybe one question to this: The compatible name for the mfd driver currently is `x-powers,axp192`. I aligned on the npm1300 implemented. But I wasn't sure if adding the driver type is the more consistant naming scheme `x-powers,axp192_mfd`.
@aasinclair  Any thoughts?